### PR TITLE
New Features and Additions

### PR DESCRIPTION
**Pull request by jxb5151**
Greetings! I'd like to fold in the following new features and improvements :) 

_Improvements:_
- Scanning: Optimized the discovery class using asyncio modules and generators for improved performance. Explorer scripts needed to be modified as well to accommodate these changes. 
- cobaltstrike_payload - Fix bugs with payload validation and better logic to detect potential Beacon payloads
- BinDataHelper - Fixed bug for null value processing and provided improved error reporting/handling. For ADD/SUB/XOR scripts introduced the `--skip-strict` option to prevent key modification (incrementing or next byte) if desired for skip null / skip key.
- Changed default output from MD5 to SHA256 for several utility scripts.
- Added JSON output feature for pepdb, peresources. 
- Updated pepdb to work with latest version of pefile. 
- Fixed a bug with 'brief' output option in petimestamp.

_Additions:_
- Added an explorer for CS beacons. See the following resources for methodology:
  - https://msrc.microsoft.com/blog/2022/10/hunting-for-cobalt-strike-mining-and-plotting-for-fun-and-profit/
  - https://github.com/splunk/melting-cobalt
- Added explorer for ShadowPad. Method discussed here:
  - https://www.youtube.com/watch?v=YCwyc6SctYs&t=1360s&ab_channel=KasperskyTech
- Added dumper for Apollo - part of a common red teaming utility.
  - https://github.com/MythicAgents/Apollo
- Added dumper for certain crypted PlugX samples (rzstreet_dumper).
  - https://blog.xorhex.com/blog/mustangpandaplugx-1/
- Added XOR brute tool to process candidate and attempt to brute force an XOR key using hamming distance, coincidence index, and known plaintext (see crypt_plaintexts).
- Added guid_recovery script for forensic recovery of .NET GUID values within a binary stream (like a mem dump).
- Added pecarver which checks for embedded PE files within a candidate and writes the results to disk. 
- Added hashes/reverse_bytes which are essentially simple convenience scripts.

### CODE REVIEW

**Comment by malwarefrank** [comment]
Code looks ok.  Maybe just update the `__version__` strings in the modules you modified.

File: malchive/active_discovery/meterpreter_reverse_shell.py [hunk 1]
```diff
@@ -13,61 +13,81 @@
...
 import struct
 import hashlib
-import requests
 from malchive.helpers import discovery
 from string import ascii_letters, digits
-from random import sample, random
+from random import sample, randint
+from async_timeout import timeout
 
 __version__ = "1.0.0"
```
Comment (malwarefrank) [inline malchive/active_discovery/meterpreter_reverse_shell.py]: if backwards-compatible changes, then increment minor component.  If breaking changes then increment major component.

File: malchive/active_discovery/spivy.py [hunk 2]
```diff
@@ -13,14 +13,14 @@
...
 import logging
 import struct
-import camellia
-import argparse
 from malchive.helpers import discovery
 from random import choice, randint
 from string import digits
+from async_timeout import timeout
 
 __version__ = "1.0.0"
```
Comment (malwarefrank) [inline malchive/active_discovery/spivy.py]: if backwards-compatible changes, then increment minor component.  If breaking changes then increment major component.

File: malchive/decoders/cobaltstrike_payload.py [hunk 3]
```diff
@@ -13,11 +13,11 @@
...
 # limitations under the License.
+#
 
 import argparse
 import logging
-import binascii
-from struct import unpack
+import struct
 
 __version__ = "1.0.0"
```
Comment (malwarefrank) [inline malchive/decoders/cobaltstrike_payload.py]: if backwards-compatible changes, then increment minor component.  If breaking changes then increment major component.

File: malchive/helpers/discovery.py [hunk 4]
```diff
@@ -13,13 +13,15 @@
...
 import socket
 import logging
 import argparse
-import ipaddress
+import netaddr
 from typing import List
 from validators import domain as valid_domain
+from async_timeout import timeout
 
 __version__ = "1.0.0"
```
Comment (malwarefrank) [inline malchive/helpers/discovery.py]: if backwards-compatible changes, then increment minor component.  If breaking changes then increment major component.

**Comment by r0ny123** [comment]
First of all, thanks for those cool features. I was testing this patch and found that there was an installation problem with `pysqlite3 0.5.0`. So, I was thinking if we could add some workflow to make sure malchive installs without an error for different os.

**Comment by jxb5151** [comment]
The contribution here hasn't modified any of the existing `pysqlite3` requirements in `setup.py` so I am not sure the comment fits in with the current PR. However, if you are experiencing issues more broadly it might be best to spin up a separate issue with more details as to the exact problem and OS and we can work through that!

**Comment by r0ny123** [comment]
Yeah, you are right. The issue doesn't fit here. I will open a issue with more details. Thanks.

**Comment by r0ny123** [comment]
@jxb5151 I opened an issue.

**Pull request merged**